File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;

public class Main {

    static int N;
    static int[][] map;
    static boolean[][] visited;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine().trim());

        map = new int[N][N];
        visited = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            String line = br.readLine().trim();
            for (int j = 0; j < N; j++) {
                map[i][j] = line.charAt(j) - '0';
            }
        }

        List<Integer> complexes = new ArrayList<>();

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (map[i][j] == 1 && !visited[i][j]) {
                    complexes.add(bfs(i, j));
                }
            }
        }

        Collections.sort(complexes);

        StringBuilder sb = new StringBuilder();
        sb.append(complexes.size()).append('\n');
        for (int count : complexes) {
            sb.append(count).append('\n');
        }

        System.out.print(sb);
    }

    static int bfs(int startX, int startY) {
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{startX, startY});
        visited[startX][startY] = true;
        int count = 0;

        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            count++;

            for (int d = 0; d < 4; d++) {
                int nx = cur[0] + dx[d];
                int ny = cur[1] + dy[d];

                if (nx >= 0 && nx < N && ny >= 0 && ny < N
                        && !visited[nx][ny] && map[nx][ny] == 1) {
                    visited[nx][ny] = true;
                    queue.offer(new int[]{nx, ny});
                }
            }
        }

        return count;
    }
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.kt
import java.util.*

fun main() {
    val n = readLine()!!.toInt()
    val map = Array(n) { readLine()!!.toCharArray().map { it - '0' }.toIntArray() }
    val visited = Array(n) { BooleanArray(n) }
    val dx = arrayOf(-1, 1, 0, 0)
    val dy = arrayOf(0, 0, -1, 1)
    val result = mutableListOf<Int>()

    fun dfs(x: Int, y: Int): Int {
        visited[x][y] = true
        var count = 1
        for (i in 0 until 4) {
            val nx = x + dx[i]
            val ny = y + dy[i]
            if (nx in 0 until n && ny in 0 until n && map[nx][ny] == 1 && !visited[nx][ny]) {
                count += dfs(nx, ny)
            }
        }
        return count
    }

    for (i in 0 until n) {
        for (j in 0 until n) {
            if (map[i][j] == 1 && !visited[i][j]) {
                result.add(dfs(i, j))
            }
        }
    }

    println(result.size)
    result.sorted().forEach { println(it) }
}